### DISCUSSION TITLE
Should we remove local, nonsparse AD indexing?

### DISCUSSION THREAD

**Discussion by lindsayad** [idaholab/moose#23308] [Polls]
There are documented cases where the local indexing, nonsparse container is faster than our default global indexing, sparse container (see https://github.com/idaholab/moose/issues/17674). However, the local indexing container is essentially unusable for MOOSE FV (at least as it is currently implemented) and generally is untenable as the element coupling stencil increases; for standard DG it is fine, but for any more element neighbor layers than 1, things are tough. Additionally the global indexing container is much easier to use as a developer.

**Comment by zachmprince** [comment]
Will removing this capability prevent us from using AD for other purposes? For instance, our optimization work is looking into using AD to compute residual derivatives w.r.t. parameter values.

**Reply by lindsayad** [reply]
No. You would likely want to use the sparse (default) container for that

**Reply by hugary1995** [reply]
I would be very surprised if forward AD (regardless of container type) works well for parameter partials... say you have one objective function and millions of quadrature points each having a parameter.

**Reply by dschwen** [reply]
Yeah, the application is not that. It would be a few model parameters or a reduced resolution parameter mesh.

**Reply by lindsayad** [reply]
We're getting off-topic but in the conversations I've heard about this, you want to get an idea about how a parameter may effect something like a postprocessor value. Forward AD could pretty easily calculate the explicit dependence of the postprocessor on a parameter, but if the postprocessor is a function of the nonlinear degrees of freedom (or even the aux system dofs) and the parameter influences the nonlinear solution (say its viscosity in a navier-stokes simulation), how do you capture the implicit dependence of the postprocessor on the parameter through the system dofs? @grmnptr I know we've touched on this stuff briefly

**Reply by zachmprince** [reply]
I think it's a little simpler than what you are describing @lindsayad . I just need the derivative of the residual w.r.t. to a parameter. So if you had N dofs and P parameters, the derivative would be a matrix N-by-P.
